lucene/analysis/icu/src/tools/java/org/apache/lucene/analysis/icu/GenerateUTR30DataFiles.java

```diff
@@ -0,0 +1,274 @@
+    final InputStreamReader reader = new InputStreamReader(stream, StandardCharsets.UTF_8);
+    final BufferedReader bufferedReader = new BufferedReader(reader);
+    StringBuilder builder = new StringBuilder();
+    String line;
+    boolean verbatim = false;
+    boolean modified = false;
+    int lineNum = 0;
+    try {
+      while (null != (line = bufferedReader.readLine())) {
+        ++lineNum;
+        if (VERBATIM_RULE_LINE_PATTERN.matcher(line).matches()) {
+          verbatim = true;
+          builder.append(line).append("\n");
+        } else {
+          Matcher ruleMatcher = RULE_LINE_PATTERN.matcher(line);
+          if (ruleMatcher.matches()) {
+            verbatim = false;
+            builder.append(line).append("\n");
+            try {
+              String leftHandSide = ruleMatcher.group(1).trim();
+              String rightHandSide = ruleMatcher.group(2).trim();
+              expandSingleRule(builder, leftHandSide, rightHandSide);
+            } catch (IllegalArgumentException e) {
+              System.err.println
+                  ("ERROR in " + file.getName() + " line #" + lineNum + ":");
+              e.printStackTrace(System.err);
+              System.exit(1);
+            }
+            modified = true;
+          } else {
+            if (BLANK_OR_COMMENT_LINE_PATTERN.matcher(line).matches()) {
+              builder.append(line).append("\n");
+            } else {
+              if (verbatim) {
+                builder.append(line).append("\n");
+              } else {
+                modified = true;
+              }
+            }
+          }
+        }
+      }
+    } finally {
+      bufferedReader.close();
+    }
+    if (modified) {
+      System.err.println("Expanding rules in and overwriting " + file.getName());
+      final FileOutputStream out = new FileOutputStream(file, false);
+      Writer writer = new OutputStreamWriter(out, StandardCharsets.UTF_8);
+      try {
+        writer.write(builder.toString());
+      } finally {
+        writer.close();
+      }
+    }
+  }
+
+  private static void getNFKCDataFilesFromIcuProject() throws IOException {
+    URL icuTagsURL = new URL(ICU_GIT_TAG_URL + "/");
+    URL icuReleaseTagURL = new URL(icuTagsURL, ICU_RELEASE_TAG + "/");
+    URL norm2url = new URL(icuReleaseTagURL, ICU_DATA_NORM2_PATH + "/");
+
+    System.err.print("Downloading " + NFKC_TXT + " ... ");
+    download(new URL(norm2url, NFKC_TXT), NFKC_TXT);
+    System.err.println("done.");
+    System.err.print("Downloading " + NFKC_CF_TXT + " ... ");
+    download(new URL(norm2url, NFKC_CF_TXT), NFKC_CF_TXT);
+    System.err.println("done.");
+
+    System.err.print("Downloading " + NFKC_CF_TXT + " and making diacritic rules one-way ... ");
+    URLConnection connection = openConnection(new URL(norm2url, NFC_TXT));
+    BufferedReader reader = new BufferedReader
+        (new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *getInputStream*. The resource is referenced in statements at the following lines: 180, 181, 186. The resource closure statement is at line: 217. There are other execution paths that do not contain closure statements, e.g., when exception is thrown by FileOutputStream constructor. Either a) close the object returned by *getInputStream()* in a try-finally block or b) close the resource by declaring the object returned by *getInputStream()* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).